Add support for the Component Model Async ABI

dicej opened the pull request:
Add support for the Component Model [Async
ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Async.md) to `wasm-tools`, `wit-bindgen`, and `wasmtime`.

[Rendered RFC](https://github.com/dicej/rfcs/blob/component-async/accepted/component-model-async.md)

tschneidereit commented on accepted/component-model-async.md:
How (un)likely are scenarios in which all imports, but only some exports should be async, or vice-versa? If they might be reasonably likely, it might make sense to support `All` or `None` for one, and `Some` for the other. I assume you have a much better intuition on this than I do, though

tschneidereit commented on accepted/component-model-async.md:
I think to some degree we don't really need this kind of discussion: the "what else could we do" question has been answered at the spec level, and we'd have had a ton of opportunity to voice disagreement there. At this point, I think we should handle this the same as other features, such as the recent support for [Tail Calls](https://github.com/bytecodealliance/rfcs/blob/main/accepted/tail-calls.md#rationale-and-alternatives).

As such, I wonder if there are possible alternative approaches to implementing support of the spec feature that would be worth discussing here? And if not, is there a way to explain why that is the case as part of a "there really isn't much of an alternative" paragraph?

tschneidereit reviewed (commented):
Hooray! 🥳

From a brief skim, the parts of this that I'm qualified to have an opinion on all look good to me. I left a couple of inline comments on specific aspects, but they're not particularly important.

dicej commented:
Ugh, I definitely did _not_ mean to merge this.  I must have pushed to origin/main by mistake.  Sorry; will revert.

dicej commented:
I've reverted all the earlier commits, rebased, and opened https://github.com/bytecodealliance/rfcs/pull/38 -- please direct further discussion there.  Sorry again for the confusion!

BTW, would it make sense to tweak the branch protection rules in this repo to block pushing straight to main?  Asking for a friend.
